src/kernel/mempool_persist.cpp

```diff
@@ -213,6 +218,7 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
                   fs::file_size(dump_path));
     } catch (const std::exception& e) {
         LogInfo("Failed to dump mempool: %s. Continuing anyway.\n", e.what());
+        (void)file.fclose();
```

Comment:
It is best if all users of `AutoFile` explicitly call the `fclose()` method and check that it succeeds (option 3. from PR description). This is not feasible because there are many users of `AutoFile`. However, some of those users already do call the `fclose()` method explicitly. For those users it is easy and I added checks whether `fclose()` succeeds. `DumpMempool()` is such a user and the change is:

```diff
             throw std::runtime_error("FileCommit failed");
-        file.fclose();           
+        if (file.fclose() != 0) {    
+            throw std::runtime_error(
+                strprintf("Error closing %s: %s", fs::PathToString(file_fspath), SysErrorString(errno)));
+        }
         if (!RenameOver(dump_path + ".new", dump_path)) {
```

For consistency in `DumpMempool()` I changed all code paths in that function to explicitly call `fclose()`. There are two codepaths and this added `(void)file.fclose();` is the other one. I ignore the return value because this is already failure handling and the function is going to return `false`. I didn't want to do something like "while handling error1, error2 occurred".

Reply:
Resolving this discussion. Changed the semantic of `AutoFile` to require all write users to call `fclose()`.